### ISSUE TITLE
Override Http DependencyTelemetry "name" Property

### ISSUE THREAD

**Issue by michaelwstark** [Microsoft/ApplicationInsights-dotnet-server#820]
I'm using the Application Insights DependencyCollector for collecting information our my service dependencies.  I wanted to be able to set the "name" property on these events similar to how I can for incoming request telemetry in order to categorize the calls without the parameters using the relative path uri templates in my source.  I've looked over the Activity class mentioned in some other threads, and it looks like I could add it if I wanted, but I saw some additional benefit in overriding the value so that each call wouldn't be unique in other areas of the Azure Portal.

Is this possible today without activities to override the name field?

My core concern is being able to slice and dice the data better, and the name field with the parameter values detailed within in the relative path make it difficult to out of the box.

**Comment by michaelwstark** [comment]
I considered a TelemetryInitializer -- that is how we achieve the same for incoming requests.  However, I don't seem to have access to the request context for the outgoing http request, and can't seem to find an easy way to grab it -- putting in on a thread context seems dangerous.  Do you have any suggestions for how I can easily get the relative url of the outgoing request?

**Comment by lmolkova** [comment]
@michaelwstark 

I think this is related to the general [issue of enriching dependency telemetry based on http request/response](https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587#issuecomment-364272438) you've commented before

Perhaps there is a bit hacky way to achieve exactly what you want with `TelemetryIntitializer`.
Raw URL is stored in  `DependencyTelmetry.Data` -  you may access it in the initializer, parse and update telemetry name.
`Name` property already has verb and relative path
`Target` represent a host name
[docs](https://docs.microsoft.com/en-us/azure/application-insights/application-insights-data-model-dependency-telemetry)
![image](https://user-images.githubusercontent.com/2347409/36332753-51eba67e-1328-11e8-81e1-c2f3b019d943.png)

More advanced scenario, when you want to use something from HTTP request that is not in the DependencyTelemetry instance, this would not work. However you may access HTTP *request* properties (not the response) by subscribing to the HTTP diagnostics source, in Start event you receive raw request and may store some properties in Activity or your custom AsyncLocal. You may use it safely in the TelemetryInititlaizer. This approach is briefly described [here](https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587#issuecomment-364272438)

The most complicated scenario is when you want to access response and enrich telemetry based on it. you can still use diagnostic source Stop event, however, this becomes hacky because AppInsights listens to the same even and your listener needs to receive the Stop event before AppInsights.

Also, as we previously discussed, we'd be happy if you can contribute to AppInsights to improve this scenario :)

But if you are interested in implementing DiagnosticSource listener - let me know I can give more info on this.

**Comment by michaelwstark** [comment]
@lmolkova Thanks for the response.  Yes, I did see the parallels with the other issue that I had commented on -- however, since that specifically dealt with headers on a response, I was unsure this should be tracked separately and didn't want to assume they could be solved generically together until we had settled on a design.  I do have some general consent on my side to contribute the headers work next iteration (March), so I should be reaching out in two weeks or so if planning goes well. If that solution also addresses this that would be a bonus.  :-)

**Comment by MS-TimothyMothra** [comment]
Hi @michaelwstark have you had the opportunity to look at this?

**Comment by michaelwstark** [comment]
Apologies for the late reply; I was on vacation last week.  Yes.  I am in the process of adding validation I synced with @lmolkova on the 12th of April.  I will be submitting a pull request no later than Wednesday the 25th.

**Comment by lmolkova** [comment]
This is implemented in #900 

Basically during TelemetryIntitialiers run you have access to HttpRequestMessage/HttpWebRequest and HttpResponseMessage/HttpWebResponse through OperationDetails:

`dependencyTelemetry.TryGetOperationDetail("HttpResponse", out var reponseObject)`

Http keys are:

```csharp
        public const string HttpRequestOperationDetailName = "HttpRequest";
        public const string HttpResponseOperationDetailName = "HttpResponse";
        public const string HttpResponseHeadersOperationDetailName = "HttpResponseHeaders";

```

Be careful: in some cases instead of HttpResponse we can only give you HttpResponseHeaders,
